src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/TriggerHelperBindingProvider.cs

```diff
@@ -0,0 +1,183 @@
+                                return o2;
+                            };
+                        }
+                    }
+                }
+
+
+                if (converter == null)
+                { 
+                    return null;
+                }
+
+                var parameter = context.Parameter;
+                var attributeSource = TypeUtility.GetResolvedAttribute<TAttribute>(parameter);
+                
+                return new ExactBinding<TUserType>
+                {
+                    _converter = converter,
+                    _directInvoker = GetDirectInvoker(cm),
+                    _getInvokeString = cm.GetConverter<TTriggerValue, DirectInvokeString, TAttribute>(),
+                    _attribute = attributeSource
+                };
+            }
+
+            private static FuncAsyncConverter<DirectInvokeString, TTriggerValue> GetDirectInvoker(IConverterManager cm)
+            {
+                var direct = cm.GetConverter<DirectInvokeString, TTriggerValue, TAttribute>();
+                if (direct != null)
+                {
+                    return direct;
+                }
+
+                var str = cm.GetConverter<string, TTriggerValue, TAttribute>();
+                if (str != null)
+                {
+                    return (input, attr, ctx) => str(input.Value, attr, ctx);
+                }
+                return null;
+            }
+
+            public async Task<IValueProvider> BindAsync(object value, ValueBindingContext context)
+            {                
+                TTriggerValue val = value as TTriggerValue;
+                if (val == null)
+                {
+                    if (_directInvoker != null && (value is string str))
+                    {
+                        // Direct invoke case. Need to converrt String-->TTriggerValue. 
+                        val = await _directInvoker(DirectInvokeString.New(str), _attribute, context);
+                    }
+                    else
+                    {
+                        // How is this possible?
+                        throw new NotImplementedException();
+                    }
+                }
+
+                TUserType result = await _converter(val, _attribute, context);
+
+                DirectInvokeString invokeString = await _getInvokeString(val, _attribute, context);
+                IValueProvider vp = new ConstantValueProvider(result, typeof(TUserType), invokeString.Value);
+                return vp;
+            }
+
+            public Task<IValueProvider> BindAsync(BindingContext context)
+            {
+                // Never called, since a trigger alreayd has an object. 
+                throw new NotImplementedException();
+            }
+
+            public ParameterDescriptor ToParameterDescriptor()
+            {
+                throw new NotImplementedException();
+            }
+
+            // Caller already filterered on TAttribute
```

Comment:
what are these floating detached comments?

Reply:
copy& paste error. will fix. 

---
In reply to: [150927366](https://github.com/Azure/azure-webjobs-sdk/pull/1422#discussion_r150927366) [](ancestors = 150927366)